pandas/core/util/hashing.py

```diff
@@ -306,8 +307,15 @@ def hash_array(
             vals = hashing.hash_object_array(vals, hash_key, encoding)
         except TypeError:
             # we have mixed types
+            try:
+                str_vals = vals.astype(str)
+            except ValueError:
+                # GH#28969 if we contain tuples, astype fails here,
+                #  apparently related to github.com/numpy/numpy/issues/9441
+                str_vals = np.array([str(x) for x in vals], dtype=object)
```

Comment:
Hmm not sure about this - wouldn't this really make everything hashable at this point, including non-hashable items?

```python
>>> vals = np.array([("a",), ["c", "d"]])
>>> str_vals = np.array([str(x) for x in vals], dtype=object)
array(["('a',)", "['c', 'd']"], dtype=object)
```

Reply:
good point